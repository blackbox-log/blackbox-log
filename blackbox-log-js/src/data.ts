import { Memoize as memoize } from 'typescript-memoize';

import { WasmPointer } from './wasm';

import type { Headers } from './headers';
import type { WasmExports, WasmObject } from './wasm';

export class DataParser implements WasmObject {
	readonly #wasm: WasmExports;
	readonly #ptr: WasmPointer;
	readonly #headers: Headers;

	constructor(wasm: WasmExports, headers: Headers, ptr: number) {
		this.#wasm = wasm;
		this.#headers = headers;
		this.#ptr = new WasmPointer(ptr, wasm.data_free);
	}

	free() {
		this.#ptr.free();
	}

	get isAlive(): boolean {
		return this.#ptr.isAlive;
	}

	get headers(): Headers {
		return this.#headers;
	}

	parseAll() {
		this.#wasm.data_parseAll(this.#ptr.ptr);
	}

	get mainFrameCount(): number {
		return this.#wasm.data_mainFrameCount(this.#ptr.ptr);
	}

	get gpsFrameCount(): number {
		return this.#wasm.data_gpsFrameCount(this.#ptr.ptr);
	}
}
