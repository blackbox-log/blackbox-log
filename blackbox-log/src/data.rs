//! Types for the data section of blackbox logs.

use alloc::vec::Vec;

use crate::event::{self, Event};
use crate::frame::{DataFrameKind, FrameKind, GpsFrame, GpsHomeFrame, MainFrame, SlowFrame};
use crate::parser::InternalError;
use crate::{Headers, Reader};

#[derive(Debug, Clone)]
pub(crate) struct Data {
    pub(crate) events: Vec<Event>,
    pub(crate) main_frames: Vec<FrameSync>,
    pub(crate) slow_frames: Vec<SlowFrame>,
    pub(crate) gps_frames: Vec<GpsFrame>,
    pub(crate) gps_home_frames: Vec<GpsHomeFrame>,
}

#[derive(Debug, Clone)]
pub(crate) struct FrameSync {
    pub(crate) main: MainFrame,
    pub(crate) slow: usize,
    pub(crate) gps: Option<usize>,
}

impl FrameSync {
    pub(crate) const fn new(main: MainFrame, slow: &[SlowFrame], gps: &[GpsFrame]) -> Self {
        Self {
            main,
            slow: slow.len() - 1,
            gps: gps.len().checked_sub(1),
        }
    }
}

/// Statistics about a decoded log.
#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
#[non_exhaustive]
pub struct Stats {
    /// The number of valid frames found of each type.
    pub counts: FrameCounts,
}

#[derive(Debug, Clone)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
pub struct FrameCounts {
    pub main: usize,
    pub slow: usize,
    pub gps: usize,
    pub gps_home: usize,
}

impl Data {
    pub(crate) fn to_stats(&self) -> Stats {
        Stats {
            counts: FrameCounts {
                main: self.main_frames.len(),
                slow: self.slow_frames.len(),
                gps: self.gps_frames.len(),
                gps_home: self.gps_home_frames.len(),
            },
        }
    }

    pub(crate) fn parse(data: &mut Reader, headers: &Headers) -> Self {
        let mut events = Vec::new();
        let mut main_frames = Vec::new();
        let mut slow_frames = Vec::new();
        let mut gps_frames = Vec::new();
        let mut gps_home_frames = Vec::new();

        slow_frames.push(headers.slow_frames.empty_frame());

        if let Some(def) = &headers.gps_frames {
            gps_frames.push(def.empty_frame());
        }

        let mut restore;
        let mut last_kind = None;
        while let Some(byte) = data.read_u8() {
            restore = data.get_restore_point();

            let Some(kind) = FrameKind::from_byte(byte) else {
                tracing::debug!("found invalid frame byte: {byte:0>#2x}");

                if let Some(last_kind) = last_kind.take() {
                    data.restore(restore);

                    match last_kind {
                        FrameKind::Event => {
                            events.pop();
                        }
                        FrameKind::Data(DataFrameKind::Intra | DataFrameKind::Inter) => {
                            main_frames.pop();
                        }
                        FrameKind::Data(DataFrameKind::Slow) => {
                            slow_frames.pop();
                        }
                        FrameKind::Data(DataFrameKind::Gps) => {
                            gps_frames.pop();
                        }
                        FrameKind::Data(DataFrameKind::GpsHome) => {
                            gps_home_frames.pop();
                        }
                    };
                }

                skip_to_frame(data);
                continue;
            };

            tracing::trace!("trying to parse {kind:?} frame");

            let result = match kind {
                FrameKind::Event => match Event::parse_into(data, &mut events) {
                    Ok(event::EventKind::End) => {
                        tracing::trace!("found the end event");
                        break;
                    }
                    Ok(_) => Ok(()),
                    Err(err) => Err(err),
                },
                FrameKind::Data(DataFrameKind::Intra | DataFrameKind::Inter) => {
                    let frame = MainFrame::parse(data, kind, &main_frames, headers);
                    frame.map(|frame| {
                        main_frames.push(FrameSync::new(frame, &slow_frames, &gps_frames));
                    })
                }
                FrameKind::Data(DataFrameKind::Gps) => headers.gps_frames.as_ref().map_or_else(
                    || {
                        tracing::debug!("found GPS frame without GPS frame definition");
                        Err(InternalError::Retry)
                    },
                    |gps| {
                        gps.parse(
                            data,
                            headers,
                            main_frames.last().map(|sync| &sync.main),
                            gps_home_frames.last(),
                        )
                        .map(|frame| gps_frames.push(frame))
                    },
                ),
                FrameKind::Data(DataFrameKind::GpsHome) => {
                    headers.gps_home_frames.as_ref().map_or_else(
                        || {
                            tracing::debug!(
                                "found GPS home frame without GPS home frame definition"
                            );
                            Err(InternalError::Retry)
                        },
                        |gps_home| {
                            gps_home
                                .parse(data, headers)
                                .map(|frame| gps_home_frames.push(frame))
                        },
                    )
                }
                FrameKind::Data(DataFrameKind::Slow) => headers
                    .slow_frames
                    .parse(data, headers)
                    .map(|frame| slow_frames.push(frame)),
            };

            match result {
                Ok(()) => {
                    last_kind = Some(kind);
                }
                Err(InternalError::Retry) => {
                    tracing::debug!("found corrupted {kind:?} frame");
                    data.restore(restore);
                    skip_to_frame(data);
                }
                Err(InternalError::Eof) => {
                    tracing::debug!("found unexpected end of file in data section");
                    break;
                }
            }
        }

        Self {
            events,
            main_frames,
            slow_frames,
            gps_frames,
            gps_home_frames,
        }
    }
}

#[cold]
fn skip_to_frame(data: &mut Reader) {
    data.skip_until_any(
        &[
            FrameKind::Event,
            FrameKind::Data(DataFrameKind::Intra),
            FrameKind::Data(DataFrameKind::Slow),
            FrameKind::Data(DataFrameKind::Gps),
            FrameKind::Data(DataFrameKind::GpsHome),
        ]
        .map(u8::from),
    );
}
