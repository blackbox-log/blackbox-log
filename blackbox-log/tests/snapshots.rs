use std::collections::BTreeMap;
use std::fs;
use std::io::Read;

use blackbox_log::data::{ParseEvent, Stats};
use blackbox_log::event::Event;
use blackbox_log::units::{si, Flag, FlagSet};
use blackbox_log::{DataParser, Headers, Unit, Value};
use serde::ser::{SerializeStruct, Serializer};
use serde::Serialize;

macro_rules! run {
    () => {
        |path| {
            let mut file = fs::File::open(path).unwrap();
            let mut data = Vec::new();
            file.read_to_end(&mut data).unwrap();

            let file = blackbox_log::File::new(&data);
            let logs = file
                .iter()
                .map(|mut reader| {
                    Headers::parse(&mut reader).map(|headers| {
                        let data = DataParser::new(&mut reader, &headers);
                        LogSnapshot::new(&headers, data)
                    })
                })
                .collect::<Vec<_>>();

            insta::assert_ron_snapshot!(logs);
        }
    };
}

#[test]
fn own() {
    insta::glob!("logs/*.bbl", run!());
}

#[test]
#[ignore]
fn fc_blackbox() {
    insta::glob!("logs/fc-blackbox/*", run!());
}

#[test]
#[ignore]
fn gimbal_ghost() {
    insta::glob!("logs/gimbal-ghost/*", run!());
}

#[derive(Debug, Serialize)]
struct LogSnapshot<'data> {
    headers: Headers<'data>,
    stats: Stats,
    events: Vec<Event>,
    main: Fields,
    slow: Fields,
    gps: Fields,
}

impl<'data> LogSnapshot<'data> {
    fn new(headers: &Headers<'data>, mut data: DataParser<'data, '_, '_>) -> Self {
        let headers = headers.clone();

        let mut events = Vec::new();
        let mut main = headers.main_def().iter().collect::<Fields>();
        let mut slow = headers.slow_def().iter().collect::<Fields>();
        let mut gps = headers
            .gps_def()
            .iter()
            .flat_map(|def| def.iter())
            .collect::<Fields>();

        while let Some(frame) = data.next() {
            match frame {
                ParseEvent::Event(event) => events.push(event),
                ParseEvent::Main(frame) => main.update(frame),
                ParseEvent::Slow(frame) => slow.update(frame),
                ParseEvent::Gps(frame) => gps.update(frame),
            }
        }

        Self {
            headers,
            stats: data.stats().clone(),
            events,
            main,
            slow,
            gps,
        }
    }
}

#[derive(Debug, Serialize)]
struct Fields(Vec<FieldSnapshot>);

impl Fields {
    fn update<F: blackbox_log::frame::Frame>(&mut self, frame: F) {
        for (field, value) in self.0.iter_mut().zip(frame.iter()) {
            field.update(value.into());
        }
    }
}

impl<T, U> FromIterator<(T, U)> for Fields
where
    T: Into<String>,
    U: Into<Unit>,
{
    fn from_iter<I: IntoIterator<Item = (T, U)>>(iter: I) -> Self {
        Self(
            iter.into_iter()
                .map(|(name, unit)| FieldSnapshot::new(name.into(), unit.into()))
                .collect(),
        )
    }
}

#[derive(Debug, Serialize)]
struct FieldSnapshot {
    name: String,
    unit: Unit,
    history: History,
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum History {
    Int(NumberHistory<16, i128>),
    Bool { yes: usize, no: usize },
    Flags(BTreeMap<&'static str, usize>),
}

#[derive(Debug)]
struct NumberHistory<const N: usize, T> {
    min: T,
    max: T,
    seen: Vec<T>,
    histogram: [usize; N],
}

impl FieldSnapshot {
    fn new(name: String, unit: Unit) -> Self {
        Self {
            name,
            unit,
            history: match unit {
                Unit::FrameTime
                | Unit::Amperage
                | Unit::Voltage
                | Unit::Acceleration
                | Unit::Rotation
                | Unit::GpsCoordinate
                | Unit::Altitude
                | Unit::Velocity
                | Unit::GpsHeading
                | Unit::Unitless => History::Int(NumberHistory::new()),
                Unit::Boolean => History::Bool { yes: 0, no: 0 },
                Unit::FlightMode | Unit::State | Unit::FailsafePhase => {
                    History::Flags(BTreeMap::new())
                }
            },
        }
    }

    #[allow(clippy::wildcard_enum_match_arm, clippy::cast_possible_truncation)]
    fn update(&mut self, value: Value) {
        match &mut self.history {
            History::Int(history) => history.update(match value {
                Value::FrameTime(t) => t.get::<si::time::microsecond>().round() as i128,
                Value::Amperage(a) => a.get::<si::electric_current::milliampere>().round() as i128,
                Value::Voltage(v) => v.get::<si::electric_potential::millivolt>().round() as i128,
                Value::Acceleration(a) => a
                    .get::<si::acceleration::centimeter_per_second_squared>()
                    .round() as i128,
                Value::Rotation(r) => {
                    r.get::<si::angular_velocity::degree_per_second>().round() as i128
                }
                Value::GpsCoordinate(c) => (c * 10000000.).round() as i128,
                Value::Altitude(a) => a.get::<si::length::meter>().round() as i128,
                Value::Velocity(v) => {
                    v.get::<si::velocity::centimeter_per_second>().round() as i128
                }
                Value::GpsHeading(h) => (h * 10.).round() as i128,
                Value::Unsigned(u) => u.into(),
                Value::Signed(s) => s.into(),
                _ => unreachable!(),
            }),
            History::Bool { yes, no } => match value {
                Value::Boolean(true) => *yes += 1,
                Value::Boolean(false) => *no += 1,
                _ => unreachable!(),
            },
            History::Flags(history) => {
                if let Value::FailsafePhase(phase) = value {
                    *history.entry(phase.as_name()).or_insert(0) += 1;
                } else {
                    let flags = match value {
                        Value::FlightMode(m) => m.as_names(),
                        Value::State(s) => s.as_names(),
                        _ => unreachable!(),
                    };

                    for flag in flags {
                        *history.entry(flag).or_insert(0) += 1;
                    }
                }
            }
        }
    }
}

impl<const N: usize, T> NumberHistory<N, T> {
    fn new() -> Self
    where
        T: Default,
    {
        Self {
            min: T::default(),
            max: T::default(),
            seen: Vec::new(),
            histogram: [0; N],
        }
    }

    fn update_range(&mut self, value: T)
    where
        T: PartialOrd,
    {
        if value < self.min {
            self.min = value;
        } else if value > self.max {
            self.max = value;
        }
    }

    /// Store new value sorted in `seen` if not already present
    fn update_seen(&mut self, value: T)
    where
        T: PartialOrd,
    {
        let index = self.seen.partition_point(|x| *x < value);
        if self.seen.get(index) != Some(&value) {
            self.seen.insert(index, value);
        }
    }

    /// Bucket so that sequential values go in sequential buckets, since data is
    /// usually clustered
    #[inline(always)]
    fn update_histogram(&mut self, value: u128) {
        self.histogram[(value % N as u128) as usize] += 1;
    }
}

impl<const N: usize> NumberHistory<N, i128> {
    fn update(&mut self, value: i128) {
        self.update_range(value);
        self.update_seen(value);
        self.update_histogram(value.unsigned_abs());
    }
}

impl<const N: usize> Serialize for NumberHistory<N, i128>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(self, self.min, self.max, serializer)
    }
}

impl<const N: usize> Serialize for NumberHistory<N, f64>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(
            self,
            format!("{:.2}", self.min),
            format!("{:.2}", self.max),
            serializer,
        )
    }
}

#[inline(always)]
fn serialize<const N: usize, T, S, U>(
    snapshot: &NumberHistory<N, T>,
    min: U,
    max: U,
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: Serializer,
    U: Serialize,
    [usize; N]: Serialize,
{
    let unique = snapshot.seen.len();

    let mut state = serializer.serialize_struct("FieldSnapshot", 4)?;

    state.serialize_field("min", &min)?;
    state.serialize_field("max", &max)?;
    state.serialize_field("unique", &unique)?;

    let histogram = "histogram";
    if unique > 1 {
        state.serialize_field(histogram, &snapshot.histogram)?;
    } else {
        state.skip_field(histogram)?;
    }

    state.end()
}
