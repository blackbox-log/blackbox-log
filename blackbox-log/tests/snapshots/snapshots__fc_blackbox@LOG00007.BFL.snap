---
source: blackbox-log/tests/snapshots.rs
expression: logs
input_file: blackbox-log/tests/logs/.fc-blackbox/src/test-data/LOG00007.BFL
---
[
  Ok(LogSnapshot(
    headers: Headers(
      version: V2,
      firmware_revision: "Betaflight 4.2.6 (a4b6db1e7) STM32F405",
      firmware_kind: Betaflight,
      board_info: Some("AIRB OMNIBUSF4"),
      craft_name: None,
      vbat_reference: Some(0),
      vbat_scale: Some(110),
      current_meter: None,
      acceleration_1g: Some(2048),
      gyro_scale: Some(0.017453292),
      min_throttle: Some(1070),
      motor_output_range: Some(MotorOutputRange(158, 2047)),
      unknown: {
        "dyn_notch_width_percent": "0",
        "rc_smoothing_active_cutoffs": "0, 90",
        "acc_hardware": "0",
        "deadband": "0",
        "tpa_rate": "65",
        "rc_smoothing_cutoffs": "0, 0",
        "gyro_lowpass2_type": "0",
        "rate_limits": "1998,1998,1998",
        "dterm_notch_cutoff": "0",
        "pid_process_denom": "1",
        "gyro_cal_on_first_arm": "0",
        "yaw_lowpass_hz": "0",
        "acc_limit_yaw": "0",
        "dyn_notch_max_hz": "400",
        "yaw_deadband": "0",
        "mag_hardware": "1",
        "maxthrottle": "2000",
        "vbat_pid_gain": "0",
        "I interval": "256",
        "dterm_lowpass_dyn_hz": "70,170",
        "use_integrated_yaw": "0",
        "dshot_bidir": "0",
        "ff_boost": "15",
        "gyro_sync_denom": "1",
        "rpm_notch_lpf": "150",
        "rc_smoothing_filter_type": "1, 1",
        "serialrx_provider": "2",
        "dshot_idle_value": "550",
        "features": "809500680",
        "feedforward_weight": "90,95,90",
        "d_min_advance": "20",
        "rc_smoothing_debug_axis": "0",
        "gyro_rpm_notch_q": "500",
        "rc_interpolation_channels": "2",
        "dterm_filter2_type": "0",
        "ff_interpolate_sp": "2",
        "gyro_lowpass_type": "0",
        "Log start datetime": "0000-01-01T00:00:00.000+00:00",
        "dterm_filter_type": "0",
        "vbatcellvoltage": "330,350,430",
        "anti_gravity_mode": "0",
        "rc_interpolation_interval": "19",
        "rc_smoothing_type": "1",
        "iterm_relax": "1",
        "gyro_rpm_notch_harmonics": "2",
        "rc_interpolation": "2",
        "tpa_breakpoint": "1350",
        "d_min_gain": "37",
        "dterm_lowpass_hz": "150",
        "rates": "70,70,70",
        "iterm_relax_cutoff": "15",
        "d_min": "23,25,0",
        "acc_limit": "0",
        "ff_spike_limit": "60",
        "ff_max_rate_limit": "100",
        "gyro_lowpass_hz": "200",
        "gyro_notch_hz": "0,0",
        "anti_gravity_threshold": "250",
        "acc_lpf_hz": "1000",
        "baro_hardware": "1",
        "motor_pwm_protocol": "7",
        "rc_rates": "100,100,100",
        "pidAtMinThrottle": "1",
        "pitchPID": "46,90,38",
        "magPID": "40",
        "P interval": "8",
        "pidsum_limit_yaw": "400",
        "pidsum_limit": "500",
        "rates_type": "0",
        "Firmware date": "Jan  5 2021 19:07:23",
        "dyn_notch_min_hz": "80",
        "iterm_relax_type": "1",
        "dterm_rpm_notch_min": "100",
        "gyro_lowpass_dyn_hz": "200,500",
        "motor_pwm_rate": "480",
        "thr_mid": "50",
        "dyn_notch_q": "250",
        "dterm_notch_hz": "0",
        "feedforward_transition": "0",
        "gyro_rpm_notch_min": "80",
        "rollPID": "42,85,35",
        "thr_expo": "0",
        "looptime": "125",
        "yawPID": "45,90,0",
        "levelPID": "50,50,75",
        "rc_expo": "0,0,0",
        "dterm_rpm_notch_harmonics": "0",
        "airmode_activate_throttle": "25",
        "debug_mode": "3",
        "P ratio": "32",
        "anti_gravity_gain": "3500",
        "gyro_hardware_lpf": "0",
        "dterm_lowpass2_hz": "150",
        "use_unsynced_pwm": "0",
        "rc_smoothing_rx_average": "0",
        "dterm_rpm_notch_q": "500",
        "iterm_windup": "100",
        "abs_control_gain": "0",
        "gyro_notch_cutoff": "0,0",
        "rc_smoothing_auto_factor": "10",
        "gyro_lowpass2_hz": "250",
      },
    ),
    stats: Stats(
      counts: FrameCounts(
        main: 294456,
        slow: 60,
        gps: 0,
        gps_home: 0,
      ),
    ),
    events: [],
    main: Fields([
      FieldSnapshot(
        name: "loopIteration",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 2355474,
          unique: 294452,
          histogram: (18406, 18404, 18405, 18403, 18403, 18403, 18403, 18403, 18403, 18403, 18403, 18405, 18403, 18403, 18403, 18403),
        ),
      ),
      FieldSnapshot(
        name: "time",
        unit: FrameTime,
        history: FieldSnapshot(
          min: 0,
          max: 303270414,
          unique: 294452,
          histogram: (18155, 18355, 18527, 18639, 18499, 18570, 18237, 18205, 18265, 18512, 18424, 18617, 18588, 18376, 18275, 18212),
        ),
      ),
      FieldSnapshot(
        name: "axisP[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -141,
          max: 383,
          unique: 263,
          histogram: (68953, 36796, 29302, 22640, 19279, 16587, 14451, 12639, 11511, 10752, 10100, 9672, 9256, 8444, 7409, 6665),
        ),
      ),
      FieldSnapshot(
        name: "axisP[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -2034,
          max: 262,
          unique: 402,
          histogram: (69148, 36171, 29974, 25157, 21687, 18685, 16232, 13591, 11779, 9842, 8820, 7840, 7172, 6823, 6121, 5414),
        ),
      ),
      FieldSnapshot(
        name: "axisP[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -2318,
          max: 296,
          unique: 429,
          histogram: (71648, 33044, 25463, 20737, 17738, 14685, 13565, 12319, 11663, 12386, 12097, 11650, 10880, 9477, 8638, 8466),
        ),
      ),
      FieldSnapshot(
        name: "axisI[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -43,
          max: 41,
          unique: 5,
          histogram: (294452, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "axisI[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -3250,
          max: 13,
          unique: 7,
          histogram: (294449, 1, 3, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 2, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "axisI[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 24,
          unique: 6,
          histogram: (294451, 1, 1, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 1, 0),
        ),
      ),
      FieldSnapshot(
        name: "axisD[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -22859,
          max: 525,
          unique: 249,
          histogram: (105392, 60820, 40457, 25572, 16764, 11562, 8416, 6186, 4706, 3578, 2864, 2219, 1899, 1569, 1320, 1132),
        ),
      ),
      FieldSnapshot(
        name: "axisD[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -676,
          max: 170314,
          unique: 228,
          histogram: (102984, 59669, 40103, 26608, 17844, 12458, 8654, 6454, 5009, 3815, 3097, 2322, 1856, 1424, 1205, 954),
        ),
      ),
      FieldSnapshot(
        name: "axisF[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -29,
          max: 84048,
          unique: 6,
          histogram: (294452, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 2, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "axisF[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -72010,
          max: 48,
          unique: 8,
          histogram: (294450, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "axisF[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -6475,
          max: 19664,
          unique: 4,
          histogram: (294454, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "rcCommand[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -10,
          max: 32,
          unique: 3,
          histogram: (294455, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "rcCommand[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -114,
          max: 0,
          unique: 5,
          histogram: (294452, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "rcCommand[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -3058,
          max: 13,
          unique: 5,
          histogram: (294452, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "rcCommand[3]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 1000,
          unique: 4,
          histogram: (4, 1, 0, 0, 0, 0, 0, 0, 294451, 0, 0, 0, 0, 0, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "setpoint[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -138,
          max: 18512,
          unique: 7,
          histogram: (294451, 0, 0, 0, 0, 1, 0, 0, 2, 0, 1, 0, 1, 0, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "setpoint[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 710864,
          unique: 6,
          histogram: (294454, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "setpoint[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -218,
          max: 51135336,
          unique: 7,
          histogram: (294451, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "setpoint[3]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -13768,
          max: 13438784,
          unique: 7,
          histogram: (294451, 0, 1, 0, 1, 0, 0, 0, 2, 0, 0, 0, 1, 0, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "vbatLatest",
        unit: Voltage,
        history: FieldSnapshot(
          min: 0,
          max: 42949672800,
          unique: 14,
          histogram: (52541, 0, 686, 0, 69986, 0, 39, 0, 6792, 0, 151754, 0, 58, 0, 12600, 0),
        ),
      ),
      FieldSnapshot(
        name: "gyroADC[0]",
        unit: Rotation,
        history: FieldSnapshot(
          min: -31056,
          max: 353763,
          unique: 204,
          histogram: (58472, 49402, 34578, 24624, 19416, 16116, 13734, 12420, 11182, 10234, 9429, 8308, 7428, 6840, 6545, 5728),
        ),
      ),
      FieldSnapshot(
        name: "gyroADC[1]",
        unit: Rotation,
        history: FieldSnapshot(
          min: -104967,
          max: 439,
          unique: 301,
          histogram: (60719, 51903, 38251, 28389, 22534, 17344, 13417, 11063, 9555, 8220, 6992, 6131, 5428, 5077, 4979, 4454),
        ),
      ),
      FieldSnapshot(
        name: "gyroADC[2]",
        unit: Rotation,
        history: FieldSnapshot(
          min: -68032519,
          max: 33659202,
          unique: 303,
          histogram: (66056, 47067, 30723, 22162, 16952, 14999, 13379, 12463, 11226, 10060, 9408, 8712, 8348, 7917, 7425, 7559),
        ),
      ),
      FieldSnapshot(
        name: "accSmooth[0]",
        unit: Acceleration,
        history: FieldSnapshot(
          min: -42325,
          max: 34423,
          unique: 1483,
          histogram: (16696, 16926, 19020, 23985, 23117, 19420, 16899, 16697, 17227, 17033, 20727, 18495, 17306, 17002, 17185, 16721),
        ),
      ),
      FieldSnapshot(
        name: "accSmooth[1]",
        unit: Acceleration,
        history: FieldSnapshot(
          min: -3309,
          max: 643,
          unique: 1450,
          histogram: (17274, 18203, 22504, 28707, 22889, 17445, 18259, 21030, 17740, 16390, 15385, 16887, 15693, 15862, 15439, 14749),
        ),
      ),
      FieldSnapshot(
        name: "accSmooth[2]",
        unit: Acceleration,
        history: FieldSnapshot(
          min: -312,
          max: 4401623,
          unique: 1952,
          histogram: (17690, 18135, 17277, 17320, 18083, 19481, 18100, 17676, 18916, 19907, 19356, 18737, 18490, 18262, 18284, 18742),
        ),
      ),
      FieldSnapshot(
        name: "debug[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -280,
          max: 177,
          unique: 199,
          histogram: (58472, 49403, 34579, 24622, 19415, 16116, 13734, 12420, 11183, 10234, 9429, 8308, 7428, 6840, 6545, 5728),
        ),
      ),
      FieldSnapshot(
        name: "debug[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -236286,
          max: 79952,
          unique: 304,
          histogram: (60720, 51903, 38252, 28389, 22534, 17344, 13417, 11063, 9553, 8220, 6992, 6131, 5428, 5077, 4980, 4453),
        ),
      ),
      FieldSnapshot(
        name: "debug[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: -39370,
          max: 96,
          unique: 299,
          histogram: (66056, 47067, 30722, 22162, 16952, 15000, 13379, 12463, 11227, 10060, 9408, 8712, 8348, 7917, 7425, 7558),
        ),
      ),
      FieldSnapshot(
        name: "debug[3]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 202832,
          unique: 7,
          histogram: (294453, 0, 1, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0),
        ),
      ),
      FieldSnapshot(
        name: "motor[0]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 1280,
          unique: 6,
          histogram: (1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 294452, 0),
        ),
      ),
      FieldSnapshot(
        name: "motor[1]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 4294715419,
          unique: 8,
          histogram: (1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 294452, 1),
        ),
      ),
      FieldSnapshot(
        name: "motor[2]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 4294962297,
          unique: 8,
          histogram: (2, 0, 0, 0, 0, 0, 1, 0, 0, 2, 2, 0, 0, 0, 294449, 0),
        ),
      ),
      FieldSnapshot(
        name: "motor[3]",
        unit: Unitless,
        history: FieldSnapshot(
          min: 0,
          max: 830112,
          unique: 6,
          histogram: (1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 294452, 2),
        ),
      ),
      FieldSnapshot(
        name: "flightModeFlags",
        unit: FlightMode,
        history: {
          "Airmode": 3,
          "Angle": 5,
          "Antigravity": 1,
          "Arm": 5,
          "BeepGpsCount": 2,
          "BeeperOn": 5,
          "Blackbox": 3,
          "BlackboxErase": 2,
          "Calib": 1,
          "CamStab": 5,
          "Camera1": 2,
          "Camera2": 3,
          "Camera3": 3,
          "Failsafe": 5,
          "FpvAngleMix": 3,
          "GpsRescue": 1,
          "HeadAdjust": 1,
          "HeadFree": 3,
          "Horizon": 3,
          "LedLow": 3,
          "Mag": 1,
          "Osd": 2,
          "Paralyze": 2,
          "Passthru": 5,
          "Servo1": 2,
          "Servo2": 3,
          "Servo3": 2,
          "Telemetry": 3,
          "ThreeD": 3,
          "Turtle": 3,
        },
      ),
      FieldSnapshot(
        name: "stateFlags",
        unit: State,
        history: {
          "GpsFix": 1,
          "GpsFixHome": 2,
        },
      ),
      FieldSnapshot(
        name: "failsafePhase",
        unit: FailsafePhase,
        history: {
          "Idle": 294453,
          "Unknown": 2,
        },
      ),
      FieldSnapshot(
        name: "rxSignalReceived",
        unit: Boolean,
        history: History(
          yes: 2,
          no: 294453,
        ),
      ),
      FieldSnapshot(
        name: "rxFlightChannelsValid",
        unit: Boolean,
        history: History(
          yes: 0,
          no: 294455,
        ),
      ),
    ]),
    gps: Fields([]),
  )),
]
